src/execution/sql/thread_state_container.cpp

```diff
@@ -83,35 +87,52 @@ void ThreadStateContainer::Reset(const std::size_t state_size, const ThreadState
 }
 
 byte *ThreadStateContainer::AccessCurrentThreadState() {
-  auto &tls_handle = impl_->states_.local();
-  return tls_handle.State();
+  common::SpinLatch::ScopedSpinLatch guard(&impl_->states_latch_);
+  auto iter = impl_->states_.find(std::this_thread::get_id());
+  byte *state = nullptr;
+  if (iter == impl_->states_.end()) {
+    auto *tls_handle = new TLSHandle(this);
+    impl_->states_.insert(std::make_pair(std::this_thread::get_id(), tls_handle));
+    state = tls_handle->State();
+  } else {
+    state = iter->second->State();
+  }
+  return state;
 }
 
 void ThreadStateContainer::CollectThreadLocalStates(std::vector<byte *> *container) const {
   container->clear();
+  common::SpinLatch::ScopedSpinLatch guard(&impl_->states_latch_);
   container->reserve(impl_->states_.size());
   for (auto &tls_handle : impl_->states_) {
```

Comment:
The loop variable isn't a `TLSHandle` any more. Suggest changing the name to `entry`, or using structured binding:
```
for (auto &[tid, tls_handle] : imple_->states_) {
  ...
}
```